apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/RetryTest.cs

```diff
@@ -0,0 +1,89 @@
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Apis.Requests;
+using Google.Apis.Storage.v1.Data;
+using Google.Cloud.ClientTesting;
+using System;
+using System.Linq;
+using System.Net;
+using Xunit;
+
+namespace Google.Cloud.Storage.V1.Tests;
+
+/// <summary>
+/// Retries custom RetryTimings and RetryPredicate variants complete flows.
+/// </summary>
+public class RetryTest
+{
+    [Theory]
+    [InlineData(2, 6, 2, 502, 4, 502, 504)]
+    [InlineData(2, 6, 2, 504, 4, 502, 504)]
+    [InlineData(2, 6, 2, 429, 0, 502, 504)]
+    [InlineData(2, 6, 2, 502, 0)]
+    public void CustomRetryPredicateTest(int statusCode, params int[] errorCodes)
+    {
+        RetryOptions retryOptions = new RetryOptions(
+            retryTiming: new RetryTiming(initialBackoff: TimeSpan.FromSeconds(2),
+            maxBackoff: TimeSpan.FromSeconds(6), backoffMultiplier: 2),
+            retryPredicate: RetryPredicate.FromErrorCodes(errorCodes));
+
+        AssertAttempts(retryOptions, statusCode, 1, 0, 2);
+    }
+
+    [Theory]
+    [InlineData(2, 6, 2, 4, 0, 2, 6, 12, 18)]
+    [InlineData(1, 5, 2, 4, 0, 1, 3, 7, 12)]
+    [InlineData(0, 1, 2, 0, 0)]
+    [InlineData(1, 2, 2, 2, 0, 1, 3)]
+    [InlineData(1, 4, 1, 3, 0, 1, 2, 3)]
+    public void CustomRetryTimingTest(int initialBackoff, int maxBackoff, double backoffMultiplier, int expectedTotalRetries, params int[] expectedBackOffs)
+    {
+        RetryOptions retryOptions = new RetryOptions(
+            retryTiming: new RetryTiming(initialBackoff: TimeSpan.FromSeconds(initialBackoff),
+            maxBackoff: TimeSpan.FromSeconds(maxBackoff), backoffMultiplier: backoffMultiplier),
+            retryPredicate: RetryPredicate.FromErrorCodes(502, 504));
+
+        AssertAttempts(retryOptions, 502, expectedTotalRetries, expectedBackOffs);
+    }
+
+    private static void AssertAttempts(RetryOptions retryOptions, int statusCode, int expectedTotalRetries, params int[] expectedBackOffs)
+    {
+        var messageHandler = new ReplayingMessageHandler(VersionHeaderBuilder.HeaderName);
+        var service = new FakeStorageService(messageHandler);
+        service.HttpClient.MessageHandler.GoogleApiClientHeader = "test/fake";
+        service.HttpClient.MessageHandler.NumTries = expectedTotalRetries + 1;
+
+        var request = service.Buckets.Get("bucket");
+        var client = new StorageClientImpl(service);
+        retryOptions ??= RetryOptions.IdempotentRetryOptions;
+
+        // Retries for the the failures.
+        service.ExpectRequests(request, (HttpStatusCode) statusCode, expectedTotalRetries);
+
+        // Last call is a success.
```

Comment:
You're making it so, yes. I was suggesting before that we should have tests for success cases separately.
We can merge like this, but we're still missing tests IMO. In particular, I don't see any tests that check that we get an exception if the retry predicate doesn't match.